src/tactic/move_add.lean

```diff
@@ -0,0 +1,397 @@
+parentheses).
+
+Note that the tactic `abel` already implements a very solid heuristic for normalizing terms in an
+additive commutative semigroup and produces expressions in more or less standard form.
+The scope of `move_add` is different: it is designed to make it easy to move individual terms
+around a sum.
+
+##  Future work
+
+* Add support for `neg/div/inv` in additive/multiplicative groups?
+* Customize error messages to mention `move_add/move_mul` instead of `move_op`?
+* Add different operations other than `+` and `*`?  E.g. `∪, ∩, ⊓, ⊔, ...`?
+  Should there be the desire for supporting more operations, it might make sense to extract
+  the `reflexivity <|> simp [add] <|> simp [mul]` block in `sorted_sum` to a separate tactic,
+  including all the lemmas used for the rearrangement to work.
+* Add functionality for moving terms across the two sides of an in/dis/equality.
+  E.g. it might be desirable to have `to_lhs [a]` converting `b + c = a + d` to `- a + b + c = d`.
+* Add a non-recursive version for use in `conv` mode.
+* Revise tests?
+-/
+
+namespace tactic
+
+namespace move_op
+
+/-!
+Throughout this file, `op : pexpr` denotes an arbitrary (binary) operation.  We do not use,
+but implicitly imagine, that this operation is associative, since we extract iterations of
+such operations, with complete disregard of the order in which these iterations arise.
+-/
+
+/-- `list_explicit_args f` returns a list of the explicit arguments of `f`. -/
+meta def list_explicit_args (f : expr) : tactic (list expr) :=
+tactic.fold_explicit_args f [] (λ ll e, return $ ll ++ [e])
+
+/--  `list_head_op op tt e` recurses into the expression `e` looking for first appearances of
+`op` as the head symbol of a subexpression.  Every time it finds one, it isolates it.
+Usually, `op` is a binary, associative operation.  E.g.,
+```lean
+#eval trace $ list_head_op ``((+)) tt `(3 / (2 + 4) + 2 * (0 + 2))
+-- [3 / (2 + 4) + 2 * (0 + 2), 2 + 4, 0 + 2]
+```
+
+More in detail, `list_head_op` partially traverses an expression in search for a term that is an
+iterated application of `op` and produces a list of such terms.
+In the intended application, the `bool` input is initially set to `tt`.
+The first time `list_head_op` finds an expression whose head symbol is `op`,
+`list_head_op` adds the expression to the list, and recurses inside the operands as well,
+but with the boolean set to `ff`.  This prevents partial operands of a large expression to
+appear.  Once `list_head_op` finds a term whose head symbol is not `op`,
+it reverts the boolean to `tt`, so that the recursion can isolate further sums later in the
+expression. -/
+meta def list_head_op (op : pexpr) : bool → expr → tactic (list expr)
+| bo (expr.app F b) := do
+  op ← to_expr op tt ff,
+  if F.get_app_fn.const_name = op.get_app_fn.const_name then do
+    Fargs ← list_explicit_args F,
+    ac ← (Fargs ++ [b]).mmap $ list_head_op ff,
+    if bo then return $ [F.mk_app [b]] ++ ac.join
```

Comment:
Rather than recomputing the application, you could do
```suggestion
| bo F'@(expr.app F b) := do
  op ← to_expr op tt ff,
  if F.get_app_fn.const_name = op.get_app_fn.const_name then do
    Fargs ← list_explicit_args F,
    ac ← (Fargs ++ [b]).mmap $ list_head_op ff,
    if bo then return $ [F'] ++ ac.join
```
Either way this look pretty suspicious to me; I don't think using either `list_explicit_args` or `const_name` is a particularly good idea. I'll have a closer look at this function once the other PR is merged.

Reply:
I'm glad that you picked up on this: I was actually thinking of producing a PR with just this function, since I also find it suspicious, after seeing what you said about the previous iterations of `list_summands/operands`.  In fact, in the branch with `compute_degree` I am running into an issue that might be related.  After applying `move_add` (and a bunch of other stuff), I am left with a goal where `norm_num` times out.  If, however I use `R : Type u` with an explicit `universe u`, instead of `R : Type*`, then `norm_num` is instantaneous.  Currently, it is not minimal at all, but I will keep monitoring whether changes to this function make the next one work better, and will report if there are issues.

Reply:
As for the change, I tried with this, closer to your solution for `list_operands`: what do you think?

```lean
meta def list_head_op (op : pexpr) : bool → expr → tactic (list expr)
| bo (expr.app F b) := match F with
  | (expr.app g a) := do
    op ← to_expr op tt ff,
    some _ ← try_core (unify op g) | (do Fc ← list_head_op tt F, bc ← list_head_op tt b,
                                      return $ Fc ++ bc),
    Fargs ← list_explicit_args F,
    ac ← (Fargs ++ [b]).mmap $ list_head_op ff,
    if bo then return $ [F.mk_app [b]] ++ ac.join
    else return ac.join
  | _ := do Fc ← list_head_op tt F, bc ← list_head_op tt b,
    return $ Fc ++ bc
  end
| bo (expr.lam _ _ e f) := do
  ec ← list_head_op tt e, fc ← list_head_op tt f,
  return $ ec ++ fc
| bo (expr.pi  _ _ e f) := do
  ec ← list_head_op tt e, fc ← list_head_op tt f,
  return $ ec ++ fc
| bo (expr.mvar  _ _ e) := do
  list_head_op tt e >>= return
| bo (expr.elet _ e f g) := do
  ec ← list_head_op tt e, fc ← list_head_op tt f, gc ← list_head_op tt g,
  return $ ec ++ fc ++ gc
| bo e := return []
```

Reply:
I just pushed a version where I merged master and `adomani_list_summands`.